### PR TITLE
Fix distribution creation and add modules to exported distribution

### LINKED ISSUE #181: Creating distribution of the developer version; crash because resource not found
### What you were trying to do
I wanted to package the latest development version with the 'dist' and 'distZip' Gradle targets to be able to distribute it

### What actually happened
When the game is started it will crash with the error message missing src/main/resources/icon.png.

### How to reproduce
- run the distZip gradle task
- open the <project directory>\desktop\build\distributions\app folder
- start sol.exe

My discoveries:
- I see that instead of the old version's res folder now the assets are packaged inside the jar file.
- The path src/main/resources/icon.png can mean the program is actually looking for the file instead of using the classloader?
- icon.png, debugOptions.ini, log4j.properties, module.info, prevShip.ini and settings.ini seems to be added twice to <project directory>\desktop\build\distributions\app\libs\sol.jar. I had a similar problem once where gradle's distribution task had two rules which packaged the same files. I will investigate further tomorrow.

### Log details and game version
commit 9efb5ac0cc985546ac2fc211c64dd600b4ca92bd

### PR DESCRIPTION
**Pull request by BenjaminAmos**
# Description
This pull request fixes the `gradlew dist` and `gradlew distZip` commands. It also adds support for including all the modules from the current workspace in the distribution,

The modules will be compiled into jars and copied into the `modules` directory when the distribution is created. Both the code and the assets should be contained within a single jar for each module, named after the module and its version.

Directory-based modules, such as those used for development, should still work when copied into the output modules directory. Zip archives do not work currently.

# Testing
## Testing distribution creation
- Run the command `gradlew dist`.
- A folder, named `app`, should be created within the `desktop/build/distributions/` directory. Enter this directory.
- Execute the relevant launcher (`sol.exe` on Windows, `solOSX.sh` on MacOS and `sol[64].sh` on Linux). The game should start and should be able to be played normally, without any additional crashes.

## Testing distribution zip creation
- Run the command `gradlew distZip`
- A zip archive should be created, called `DestinationSol.zip`, within the `desktop/build/distributions` directory.
- Extract `DestinationSol.zip` into a directory and enter it,
- Execute the relevant launcher (`sol.exe` on Windows, `solOSX.sh` on MacOS and `sol[64].sh` on Linux). The game should start and should be able to be played normally, without any additional crashes.

### Testing code-only modules
- Run the command `groovyw module get codeModuleTest -remote BenjaminAmos` to get my codeModuleTest module.
- Follow the steps contained within the "Testing distribution creation" section.
- Check that in the top-left corner of the screen there is some text labelled "Game Duration: ", followed by the quantity of seconds that the current game session has run for.

### Testing asset-only modules
- Follow the steps contained within the "Testing distribution creation" section.
- Check that the game does not crash when starting a new game. If the core module is not present or loaded then the game will crash.

### Testing modules containing both assets and code
- Run the command `groovyw module get warp` to get the warp module.
- Follow the steps contained within the "Testing distribution creation" section.
- Start a new game with the "Warp Scout" ship.
- Check that the game does not crash.
- Check that there is a button on the right side of the screen labelled "Research".

# Notes
- A dependency on the package `com.github.marschall:zipfilesystem-standalone` has been added to the engine module in order to fix the built-in jre distribution (see https://github.com/MovingBlocks/DestinationSol/pull/300#issuecomment-497719782 for more information).
- This has not been tested on Android.
- I have tested this on Windows and Linux (Ubuntu 19.04) but not on MacOS.
- Modules are not loaded successfully when provided as zip archives (.zip).
- On Linux, the executable permissions are broken, so you will need to add execute rights to the following files:
  - `lwjreLinux/bin/java`
  - `sol.sh`
  - `sol64.sh`
- If merged, this can be considered to supersede #300.
- This fixes #181.

# Credits
- To **Avalancs** for providing the intial changes that partially resolved the issue (see #181)
- To **NicholasBatesNZ** for providing additional improvements to **Avalancs**'s code (see #300)
- To **Adrijaned** for making clear that build.gradle should be ignored in modules (see https://github.com/MovingBlocks/DestinationSol/pull/300#issuecomment-497729991)
- I referenced Terasology's code for this and certain methods used may be similar, if not identical, to code in the following files:
  - [build.gradle](https://github.com/MovingBlocks/Terasology/blob/dbc32f0da97e46c447ae2293a00e5babaf92c85b/build.gradle#L145-L148)
  - [facades/PC/build.gradle](https://github.com/MovingBlocks/Terasology/blob/dbc32f0da97e46c447ae2293a00e5babaf92c85b/facades/PC/build.gradle#L225-L277)
  - [config/groovy/module.groovy](https://github.com/MovingBlocks/Terasology/blob/dbc32f0da97e46c447ae2293a00e5babaf92c85b/config/groovy/module.groovy#L46-L63)

### CODE REVIEW

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

File: engine/src/main/java/org/destinationsol/assets/AssetDataFileHandle.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018 MovingBlocks
```
Comment (NicholasBatesNZ) [inline engine/src/main/java/org/destinationsol/assets/AssetDataFileHandle.java]: It's 2019 :P

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by BenjaminAmos** [comment]
When I copied the zip archive produced on Windows into Linux the permissions were missing. They seem to be correct if compiling on Linux directly.

I have added some commits to reduce the logging level in distribution builds (or more precisely to increase it in IDE and gradle builds). You may need to run the commands `gradlew cleanIdea` and `gradlew idea` in order to use those changes.

I also applied the fix suggested by **Avalancs** in https://github.com/MovingBlocks/DestinationSol/issues/181#issuecomment-349786851, which appears to have fixed the duplicate files in sol.jar.

**Comment by Cervator** [comment]
Added commit for Macs is https://github.com/MovingBlocks/DestinationSol/commit/6ab488d1b0b075c59c622d0a67978944e6ed93e4 - included also making the Mac script executable. Although we might want a `.command` instead - also breaks embedded JRE for Mac which needs a refix along with #414

**Comment by Cervator** [comment]
Oh, and thanks for the great testing instructions by the way. I happened to have both your content module and test module in my workspace still and tested with both :-)

(the mega-weapon on one ship in "warp" is semi-broken by some faction changes though, i think the projectiles insta-collide with each other now)

**Pull request merged**